lib/model/rwfolder.go

```diff
@@ -1372,12 +1374,49 @@ func (f *rwFolder) dbUpdaterRoutine() {
 	tick := time.NewTicker(maxBatchTime)
 	defer tick.Stop()
 
+	var changedFiles []string
+	var changedDirs []string
+	if f.fsync {
+		changedFiles = make([]string, 0, maxBatchSize)
+		changedDirs = make([]string, 0, maxBatchSize)
+	}
+
+	syncFilesOnce := func(files []string, syncFn func(string) error) {
+		sort.Strings(files)
+		var lastFile string
+		for _, file := range files {
+			if lastFile != file {
+				lastFile = file
+				if err := syncFn(file); err != nil {
+					l.Infof("fsync %q failed: %v", file, err)
+				}
+			}
+		}
+	}
+
 	handleBatch := func() {
 		found := false
 		var lastFile protocol.FileInfo
 
 		for _, job := range batch {
 			files = append(files, job.file)
+			if f.fsync {
+				// collect changed files and dirs
+				switch job.jobType {
+				case dbUpdateShortcutFile:
```

Comment:
Metadata is **not** stored in the directory entry. fsync of the file also syncs metadata.

The directory entry only contains the filename and a pointer to the file structure. Syncing the directory is only necessary after creating, renaming or removing a file.